gcp-pubsub/src/main/java/io/micronaut/gcp/pubsub/intercept/PubSubClientIntroductionAdvice.java

```diff
@@ -0,0 +1,189 @@
+import io.micronaut.gcp.pubsub.support.PublisherFactory;
+import io.micronaut.inject.ExecutableMethod;
+import io.micronaut.messaging.annotation.Body;
+import io.micronaut.messaging.annotation.Header;
+import io.micronaut.scheduling.TaskExecutors;
+import io.reactivex.Scheduler;
+import io.reactivex.Single;
+import io.reactivex.schedulers.Schedulers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Named;
+import javax.inject.Singleton;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+
+/**
+ * Implementation of {@link io.micronaut.gcp.pubsub.annotation.PubSubClient} advice annotation.
+ *
+ * @author Vinicius Carvalho
+ * @since 2.0.0
+ */
+@Singleton
+public class PubSubClientIntroductionAdvice implements MethodInterceptor<Object, Object> {
+
+    private final Logger logger = LoggerFactory.getLogger(PubSubClientIntroductionAdvice.class);
+    private final PublisherFactory publisherFactory;
+    private final PubSubMessageSerDesRegistry serDesRegistry;
+    private final Scheduler scheduler;
+    private final ConversionService<?> conversionService;
+    private final GoogleCloudConfiguration googleCloudConfiguration;
+    private final ConcurrentHashMap<ExecutableMethod, PubSubPublisherState> publisherStateCache = new ConcurrentHashMap<>();
+
+    public PubSubClientIntroductionAdvice(PublisherFactory publisherFactory,
+                                          PubSubMessageSerDesRegistry serDesRegistry,
+                                          @Named(TaskExecutors.IO) ExecutorService executorService,
+                                          ConversionService<?> conversionService,
+                                          GoogleCloudConfiguration googleCloudConfiguration) {
+        this.publisherFactory = publisherFactory;
+        this.serDesRegistry = serDesRegistry;
+        this.scheduler = Schedulers.from(executorService);
+        this.conversionService = conversionService;
+        this.googleCloudConfiguration = googleCloudConfiguration;
+    }
+
+    @Override
+    public Object intercept(MethodInvocationContext<Object, Object> context) {
+
+        if (context.hasAnnotation(Topic.class)) {
+
+            PubSubPublisherState publisherState = publisherStateCache.computeIfAbsent(context.getExecutableMethod(), method -> {
+                AnnotationValue<PubSubClient> client = method.findAnnotation(PubSubClient.class).orElseThrow(() -> new IllegalStateException("No @PubSubClient annotation present"));
+                String projectId = client.getValue(String.class).orElse(googleCloudConfiguration.getProjectId());
+                AnnotationValue<Topic> topicAnnotation = method.findAnnotation(Topic.class).get();
+                String topic = topicAnnotation.getValue(String.class).get();
+                String contentType = topicAnnotation.get("contentType", String.class).orElse("");
```

Comment:
Use `stringValue("contentType")` instead

Reply:
@graemerocher I noticed that because this has a default value on the annotation, `stringValue("contentType")` returns empty, while `topicAnnotation.get("contentType", String.class)` returns the right value. So only this one I had to left as it is for tests to complete.